### ISSUE TITLE
Could you please explain the rationale behind including both input and output in the data format for a causal language model?

### ISSUE THREAD

**Issue by Beomi** [Sea-Snell/JAXSeq#3]
Hello,

I came across this repository while searching for ways to train a 10~20B scale causal language model (similar to GPT) using TPUv4 pod.

In the README, I noticed that the recommended data format is {'train': [{'in_text', 'out_text'}, ...], 'eval': [{'in_text', 'out_text'}, ...]}. Could you please explain the reasoning behind this format?

As far as I understand, GPT-2 or GPT-J only requires single texts as input and output for training, instead of the input-output pairs mentioned in the data format. I appreciate any clarification you can provide on this matter. 

Thank you.

**Comment by Beomi** [comment]
Hi, I wanted to inquire about a section of code in the JAXSeq repository that I found while working on my project. Specifically, I came across the following lines:

https://github.com/Sea-Snell/JAXSeq/blob/45a7a3365b9f5856000a58503ee8936c9441da44/src/seq2seq.py#L186

https://github.com/Sea-Snell/JAXSeq/blob/45a7a3365b9f5856000a58503ee8936c9441da44/src/seq2seq.py#L387

Based on my understanding, input_ids and decoder_input_ids are in_tokens and out_tokens, respectively, and are concatenated with axis=1. This seems to be similar to adding strings together using the '+' operator.

```python
# Define two 1D arrays of integers
input_ids = jnp.array([1, 2, 3])
decoder_input_ids = jnp.array([4, 5, 6])

# Concatenate the two arrays along axis 1
full_input_ids = jnp.concatenate((input_ids, decoder_input_ids,), axis=1)
# full_input_ids will be [1,2,3,4,5,6]
```

Given this, I was wondering if it would be appropriate to construct datasets for training GPT-like models (such as GPT-2 and GPT-J) in the following way:

```json
{
  "train": [
    {"in_text": "sentence to train1", "out_text": ""},
    {"in_text": "sentence to train2", "out_text": ""},
    ...
  ], 
  "eval": [
    {"in_text": "sentence to eval", "out_text": ""},
    ...
  ], 
}
```

I would appreciate any insights or guidance you can provide on this matter.

Thank you for your time and assistance.

**Comment by Sea-Snell** [comment]
Thanks for the question! First of all, often we are fine-tuning on data that has the form of (input, output), so this abstraction of input/output is generally useful. And secondly the input/output abstraction is strictly more general than just modeling sequences. Because if you are just modeling a sequence there is still usually a beginning of sequence token, which could be considered the input and the rest the output.

As for your specific example. You would want in_text to be the beginning of sequence token and out_text to be "sentence to train" because the out text gets trained on the in text doesn't.

**Comment by Beomi** [comment]
Thank you for your response! 

Your explanation on the input/output abstraction and how it is generally useful for fine-tuning on data is very helpful. 

I also appreciate your clarification on how the input/output abstraction is more general than just modeling sequences, 
and how a beginning of sequence token can be considered the input while the rest of the sequence is considered the output.
Regarding my case, I will take this into consideration for my project. 

Thank you again for your insights!
